### PR TITLE
Convert History, Saved, and Blacklist to YapDatabase

### PR DESCRIPTION
**Pull request by coreyfloyd**
Finally™

Lots of changes… obviously:
- `MWKDataStore` now has a `YapDatabase` where it stores history, saved, and blacklist data
- History, Blacklist, and Saved have been coalesced into he same object - MWKHistoryEntry. This is much more convenient to have one object instead of 3 - now you can do queries on the database (isSaved || isInHistory) or (isInHistory && !inBlackList).
- To power a list, YapDatabase works on the concept of Views. You need to assemble several components to create a view. I won't go into details here, but you can read up on it [here](https://github.com/yapstudios/YapDatabase/wiki/Views). I've tried my best to hide most of these details. Instead you can work directly with…
- WMFDatabaseDataSource - I created a new protocol call WMFDataSource. This is inspired by the conversations we have had about "what is wrong with SSDataSources". WMFDataBaaeDataSource is the concrete implementation of this which works directly with the database and hides most of the complexity from view controllers and other higher level components (like MWKHistoryList)
- Additionally, there is a MWKItemUpdatedNotification that MWKDataStore posts when any object is updated, and you can listen to this instead of using a dataSource (this is what the savedButtonController does)
- The Desktop URL is the key for all objects

Thats the high level concepts. Some things to keep in mind:
- Only history items in are currently in the DB, Articles are still on Disk. This is just a time constraint.
- 🚨 Writing the the DB is always asynchronous. You may make your changes synchronously, but the database will always be updated on a background queue. You should wait for the notification before updating the UI. WMFDatabaseDataSource takes care of this for you and allows you to animate any changes in your table/collection view. Or you can use the MWKItemUpdatedNotification to respond to changes
- The details of YapDatabase that leak through mostly have to do with Transactions. They would be very tough to hide. You can see this in the read/write methods
- MWKHistoryEntry is now poorly named - but I did not want to change the name and make this diff any bigger, that will be done as soon as this is merged. I have a straw man name below - feel free to suggest another.
- MWKArticle could use a rename as well… 😉 
- MWKList is nearly 👋 but we still need to get rid of recent searches to do that…
- Most of work that went into this PR had to do with figuring out the architecture. I've iterated about 4 times on the shape of it. Its the best I have some up with so far, but still leaks more than I would like. Any ideas are welcome.

Progress:
- [x] Convert Saved page List
- [x] Convert History List
- [x] Convert BlackList
- [x] Convert Schema manager
- [x] Convert TableViews
- [x] Convert Spotlight manger
- [x] Convert SaveButtonController
- [x] Migrate BlacklIst and Saved to sue MWKHistoryEntry (Deprecate MWKSavedPageEntry)
- [x] Test Save page Fetcher Saving
- [x] Fix Unit Tests

Follow-ons to come:
- [ ] Rename MWKHistoryEntry to WMFArticleReference
- [ ] Rename MWKArticle to WMFArticleContent
- [ ] Denormalize thumb, description, and snippet into MWKHistoryEntry
- [ ] Do not pull article into memory to populate cells in history and saved (use the denormalized data)
- [ ] Use search data to populate MWKHistoryEntry thumb, description, and snippet 
- [ ] Use article data to populate MWKHistoryEntry thumb, description, and snippet (this is needed for any saved or history that existed prior to this update)

Out of Scope:
- [ ] Store MWKArticles in YapDatabase

### CODE REVIEW

**Comment by coreyfloyd** [comment]
This is going to fail…

**Comment by joewalsh** [comment]
:+1: from the walkthrough & testing

File: Wikipedia/Code/MWKCitation.m [hunk 1]
```diff
@@ -90,6 +90,11 @@ - (NSArray*)backlinkIdentifiers {
     return _backlinkIdentifiers;
 }
 
+- (NSString*)description {
+    //Do not use MTLModel's description as it will cuase recursion since this instance has a reference to the article, which also has a reference to this citation
```
Comment (montehurd) [inline Wikipedia/Code/MWKCitation.m]: Typo

Comment (montehurd) [inline Wikipedia/Code/MWKCitation.m]: Typo on 'cuase'

File: Wikipedia/Code/MWKImage.m [hunk 2]
```diff
@@ -213,8 +213,8 @@ - (BOOL)isVariantOfImage:(MWKImage*)otherImage {
 }
 
 - (NSString*)description {
-    return [NSString stringWithFormat:@"%@ article: %@ sourceURL: %@",
-            [super description], self.article.url, self.sourceURLString];
+    //Do not use MTLModel's description as it will cuase recursion since this instance has a reference to the article, which also has a reference to this image
```
Comment (montehurd) [inline Wikipedia/Code/MWKImage.m]: Typo on 'cuase'

File: Wikipedia/Code/MWKProtectionStatus.m [hunk 3]
```diff
@@ -64,4 +64,10 @@ - (id)copyWithZone:(NSZone*)zone {
     return self;
 }
 
+- (NSString*)description {
+    //Do not use MTLModel's description as it will cuase recursion since this instance has a reference to the article, which also has a reference to this image
```
Comment (montehurd) [inline Wikipedia/Code/MWKProtectionStatus.m]: Typo on 'cuase'

File: Wikipedia/Code/MWKSection.m [hunk 4]
```diff
@@ -154,7 +154,8 @@ - (BOOL)isEqualToSection:(MWKSection*)section {
 }
 
 - (NSString*)description {
-    return [NSString stringWithFormat:@"%@ id: %d line: %@ level: %@", [super description], self.sectionId, self.line, self.level];
+    //Do not use MTLModel's description as it will cuase recursion since this instance has a reference to the article, which also has a reference to this section section
```
Comment (montehurd) [inline Wikipedia/Code/MWKSection.m]: Typo on 'cuase'

File: Wikipedia/Code/MWKSectionList.m [hunk 5]
```diff
@@ -164,10 +164,11 @@ - (void)save {
     }
 }
 
-- (NSString*)debugDescription {
-    return [NSString stringWithFormat:@"%@ { \n"
+- (NSString*)description {
+    //Do not use MTLModel's description as it will cuase recursion since this instance has a reference to the article, which also has a reference to this section list
```
Comment (montehurd) [inline Wikipedia/Code/MWKSectionList.m]: Typo on 'cuase'

File: Wikipedia/Code/SavedArticlesFetcher.m [hunk 6]
```diff
@@ -182,7 +207,7 @@ - (void)fetchArticleURL:(NSURL*)articleURL failure:(WMFErrorHandler)failure succ
 
 - (void)downloadImageDataForArticle:(MWKArticle*)article failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
     if (![[NSUserDefaults standardUserDefaults] wmf_didFinishLegacySavedArticleImageMigration]) {
-        WMF_TECH_DEBT_TODO(This legacy migration can be removed after enough users upgrade to 5.0.5)
+        WMF_TECH_DEBT_TODO(This legacy migration can be removed after enough users upgrade to 5.0 .5)
```
Comment (montehurd) [inline Wikipedia/Code/SavedArticlesFetcher.m]: Extra space in "5.0 .5" - deliberate?

File: Wikipedia/Code/YapDatabase+WMFViews.h [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+- (YapDatabaseViewGrouping*)wmf_historyGroupingUngrouped;
+- (YapDatabaseViewGrouping*)wmf_savedGroupingUngrouped;
+- (YapDatabaseViewGrouping*)wmf_historyOrSavedGroupingUngrouped;
+- (YapDatabaseViewGrouping*)wmf_historyGroupingByDate;
+
+- (YapDatabaseViewGrouping*)wmf_blackListGroupingUngrouped;
+- (YapDatabaseViewGrouping*)wmf_notInHistorySavedOrBlackListGroupingUngrouped;
+
+- (YapDatabaseViewSorting*)wmf_historySortedByDateDecending;
```
Comment (montehurd) [inline Wikipedia/Code/YapDatabase+WMFViews.h]: Typo on 'decending'

File: Wikipedia/Code/YapDatabase+WMFViews.h [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+/**
+ *  Mappings for Views where group names are stringified NSTimeIntervals 
+ *  This sorts them in decending order.
+ *
+ *  @param viewName The view to apply mappings to
+ *
+ *  @return The mappings
+ */
+- (YapDatabaseViewMappings*)wmf_groupsAsTimeIntervalsSortedDecendingMappingsWithView:(NSString*)viewName;
```
Comment (montehurd) [inline Wikipedia/Code/YapDatabase+WMFViews.h]: Typo on 'decending'.

File: Wikipedia/Code/YapDatabase+WMFViews.h [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ *  Register the views with the names included above.
+ *  If you create a new peristent view, you should register it within this method.
+ */
+- (void)wmf_registerViews;
+
+#pragma mark - Mappings
+/**
+ *  Mappings are used to sort groups in to sections fit for collection views / table views
+ *  Mappings are required for views with mutliple sections that must be displayed in a collection view or table view. They are not for views with a single section.
```
Comment (montehurd) [inline Wikipedia/Code/YapDatabase+WMFViews.h]: Typo on mutliple

File: Wikipedia/Code/YapDatabase+WMFViews.h [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+ *  Mappings are also useful for views with single groups simply to provide access via indexPath.
+ *
+ *  Although mappings are seperate objects, they are explicitly tied to a view on a instantiation
```
Comment (montehurd) [inline Wikipedia/Code/YapDatabase+WMFViews.h]: Typo on seperate

File: Wikipedia/Code/MWKDataStore.m [hunk 11]
```diff
@@ -31,54 +54,154 @@ @interface MWKDataStore ()
...
+    if ([notifications count] == 0) {
+        return;
+    }
+    
+    [self.changeHandlers compact];
+    for (id<WMFDatabaseChangeHandler> obj in self.changeHandlers){
+        [obj processChanges:notifications onConnection:self.articleReferenceReadConnection];
+    }
+    
+    //Order is simportant.
```
Comment (montehurd) [inline Wikipedia/Code/MWKDataStore.m]: Typo on simportant

**Comment by montehurd** [comment]
Perhaps instead of renaming...

     MWKArticle to WMFArticleContent
     MWKHistoryEntry to WMFArticleReference

...we could consider...

     Leaving MWKArticle as-is
     Changing MWKHistoryEntry to WMFArticleMetaData

(Yap makes a distinction between modifications made to objects vs object meta data... thought we might just go with same convention. No strong feelings one way or the other.)

**Comment by montehurd** [comment]
@coreyfloyd testing for quite a bit... wasn't able to find any issues. did u want to make those naming changes pre or post merge?

**Comment by coreyfloyd** [comment]
> (Yap makes a distinction between modifications made to objects vs object meta data... thought we might just go with same convention. No strong feelings one way or the other.)

@montehurd  Not a bad thought, and I was thinking the same thing… BUT we can't actually use them as metadata objects since metadata objects need an object to be associated with. And since we don't always have an article it won't work. 

And if we can't actually use them as metadata objects in the database sense, I thought it was not best to name them that way to eliminate confusion. 

I'm not married to what to rename things too either… @joewalsh actually came up with article reference… and I honestly couldn't think of anything better.

I do prefer to not to leave the article as-is - and would like to add some specificity to the name. Whether its ArticleContent or something else. Just something that says "hey - the stuff that makes up an article is in here, and things like saved and history status are not"

**Comment by coreyfloyd** [comment]
> @coreyfloyd testing for quite a bit... wasn't able to find any issues. did u want to make those naming changes pre or post merge?

I was going to do the typo fixes now. I can do the article and history rename separately or in here - I'm ok either way

**Comment by coreyfloyd** [comment]
> I was going to do the typo fixes now. I can do the article and history rename separately or in here - I'm ok either way

@montehurd On second thought lets do the renames after - just in case it fucks up unit tests… they are all working now. Don't want to upset the unit test gods.

**Pull request merged**